### PR TITLE
Make sure to revalidate instances when mixins are invalidated

### LINKED ISSUE #21: Mixins applied inconsistently
REF https://github.com/Polymer/polymer/issues/4115
Distributed nodes appear to be styled with mixins defined in the host, providing they are applied after the host has been defined.

Please note the following demo: http://codepen.io/oridan/pen/edqMPO?editors=1000
Item B simply extends Item A with no other changes. Therefore they should both display exactly the same. The expected result as I understand it should be per as the following image:
image

**RFC:**
I have resolved this where `nativeCssVariables == true` by applying the following two diffs:
https://github.com/webcomponents/shadycss/compare/master...TomK:apply-host-descendants - quite inefficient
https://github.com/webcomponents/shadycss/compare/master...TomK:refresh-invalid-shim

I understand this is not a complete resolution, and possibly this implementation is not feasible.  So I am hoping for comments/suggestions to steer towards a viable solution.

### PR DESCRIPTION
**Pull request by dfreedm**
Fixes #21

### CODE REVIEW

**Comment by TomK** [comment]
this works for me 👍

File: src/ShadyCSS.js [hunk 1]
```diff
@@ -150,17 +152,21 @@ export let ShadyCSS = {
     if (!styleInfo) {
       styleInfo = this._prepareHost(host);
     }
+    assign(styleInfo.overrideStyleProperties, overrideProps);
```
Comment (sorvell) [inline src/ShadyCSS.js]: remove

File: src/ShadyCSS.js [hunk 2]
```diff
@@ -180,18 +186,25 @@ export let ShadyCSS = {
...
+  _applyToDescendants(children) {
+    for (let i = 0, c; i < children.length; i++) {
+      c = children[i];
       if (c.shadowRoot) {
         this.applyStyle(c);
+      } else if (c.localName === 'slot') {
+        if (!nativeShadow) {
+          window.ShadyDOM.flush();
+        }
+        this._applyToDescendants(c.assignedNodes().filter(this._isElementNode));
```
Comment (sorvell) [inline src/ShadyCSS.js]: Let's add an inline comment about why this needs to be here in this exact case.

**Review by sorvell — CHANGES_REQUESTED** [review]

File: src/ShadyCSS.js [hunk 3]
```diff
@@ -146,52 +149,74 @@ export let ShadyCSS = {
...
+  },
+  _isElementNode(node) {
+    return node.nodeType === Node.ELEMENT_NODE;
+  },
+  _applyToDescendants(children) {
+    for (let i = 0, c; i < children.length; i++) {
+      c = children[i];
       if (c.shadowRoot) {
         this.applyStyle(c);
+      } else if (nativeCssVariables && c.localName === 'slot') {
```
Comment (sorvell) [inline src/ShadyCSS.js]: Let's consider not doing this as it's a perf hit (yes, that means the issue still exists).

File: src/ShadyCSS.js [hunk 4]
```diff
@@ -146,52 +149,74 @@ export let ShadyCSS = {
...
-      if (template && template.__applyShimInvalid && template._style) {
-        // update template
-        ApplyShim.transformRules(template._styleAst, is);
-        template._style.textContent = StyleTransformer.elementStyles(host, styleInfo.styleRules);
-        // update instance if native shadowdom
-        if (this.nativeShadow) {
-          let style = host.shadowRoot.querySelector('style');
-          style.textContent = StyleTransformer.elementStyles(host, styleInfo.styleRules);
+    let template = templateMap[is];
+    if (this._forceRecalc || styleInfo.shouldRecalculate || template._applyShimInvalid) {
```
Comment (sorvell) [inline src/ShadyCSS.js]: let's also bail out of this function early if we detect the element does not use or produce `apply` rules and native properties are in use.

**Review by sorvell — CHANGES_REQUESTED** [review]

**Pull request closed without merging**